### PR TITLE
Websockets

### PR DESCRIPTION
**Pull request by ohaddahan**

### CODE REVIEW

File: libs/extension/src/background/ws/websocket.rs [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+use super::{on_error_handler, on_message_handler, on_open_handler};
+use block_mesh_common::constants::DeviceType;
+use leptos::SignalGetUntracked;
+use logger_leptos::leptos_tracing::setup_leptos_tracing;
+use wasm_bindgen::prelude::*;
+use web_sys::WebSocket;
+
+#[wasm_bindgen]
+pub async fn start_websocket() -> Result<(), JsValue> {
```
Comment (ohaddahan) [inline libs/extension/src/background/ws/websocket.rs:11]: need reconnect on disconnect

File: libs/speed-test/src/utils/metadata.rs [hunk 2]
```diff
@@ -37,6 +37,21 @@ pub async fn fetch_metadata() -> anyhow::Result<Metadata> {
     })
 }
 
+pub fn fetch_metadata_blocking() -> anyhow::Result<Metadata> {
+    let url = &format!("{}/{}{}", BASE_URL, DOWNLOAD_URL, 0);
+    let headers = reqwest::blocking::get(url)
```
Comment (ohaddahan) [inline libs/speed-test/src/utils/metadata.rs:42]: shouldn't use blocking

File: libs/block-mesh-manager/src/ws/handle_socket.rs [hunk 3]
```diff
@@ -33,22 +33,33 @@ pub async fn handle_socket(
...
+                    tracing::trace!("WS Text: {text}");
+                    notify_r.notify_one(); // only for demo
+                    match serde_json::from_str::<WsMessage>(&text) {
+                        Ok(ws_message) => {
+                            match ws_message.message {
+                                WsMessageTypes::SendTaskFromServer(_) => {}
+                                WsMessageTypes::SubmitTaskToServer(task) => {
+                                    tokio::time::sleep(std::time::Duration::from_secs(2)).await;
+                                    // TODO: Register this task in DB as Completed
+                                    notify_r.notify_one();
```
Comment (Adimac93) [inline libs/block-mesh-manager/src/ws/handle_socket.rs:52]: Notify that user successfully sent task to the server.

File: libs/block-mesh-manager/src/ws/handle_socket.rs [hunk 4]
```diff
@@ -58,65 +69,29 @@ pub async fn handle_socket(
...
-        };
-        tokio::time::sleep(std::time::Duration::from_secs(2)).await;
-        sender
-            .send(Message::Text(serde_json::to_string(&ws_message).unwrap()))
-            .await
-            .unwrap();
-    }
+    let send_task = tokio::spawn(async move {
+        loop {
+            let task_receiver = task_manager.add_session().await;
```
Comment (Adimac93) [inline libs/block-mesh-manager/src/ws/handle_socket.rs:74]: 1. Marks session as ready in the task manager

File: libs/block-mesh-manager/src/ws/handle_socket.rs [hunk 5]
```diff
@@ -58,65 +69,29 @@ pub async fn handle_socket(
...
+            let task = task_receiver.await.unwrap(); // waits for new task
```
Comment (Adimac93) [inline libs/block-mesh-manager/src/ws/handle_socket.rs:75]: 2. Waits for task assignment

File: libs/block-mesh-manager/src/ws/handle_socket.rs [hunk 6]
```diff
@@ -58,65 +69,29 @@ pub async fn handle_socket(
...
+                email: Some(org_email.to_string()),
+                device: None,
+                message: WsMessageTypes::SendTaskFromServer(task),
+            };
+            // tokio::time::sleep(std::time::Duration::from_secs(2)).await;
+            sender
+                .send(Message::Text(serde_json::to_string(&ws_message).unwrap()))
+                .await
+                .unwrap();
+            notify.notified().await;
```
Comment (Adimac93) [inline libs/block-mesh-manager/src/ws/handle_socket.rs:87]: Blocks until the task is send to the server

File: libs/block-mesh-manager/src/ws/ws_handler.rs [hunk 7]
```diff
@@ -39,10 +39,9 @@ pub async fn ws_handler(
     //     .await
     //     .map_err(|_| Error::Auth("Can't find token".to_string()))?;
 
-    let ws_connection_manager = state.ws_connection_manager.clone();
```
Comment (Adimac93) [inline libs/block-mesh-manager/src/ws/ws_handler.rs]: Drop idea of connection manager.

File: libs/block-mesh-manager/src/worker/ws_worker.rs [hunk 8]
```diff
@@ -0,0 +1,35 @@
+use block_mesh_common::interfaces::ws_api::{WsMessage, WsMessageTypes};
```
Comment (ohaddahan) [inline libs/block-mesh-manager/src/worker/ws_worker.rs:1]: no change here? @Adimac93

File: libs/block-mesh-manager/src/ws/connection_manager.rs [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+    }
+    pub fn register(&self, user_id: Uuid, transmitter: SplitSink<WebSocket, Message>) {
+        if let Some(old_value) = self.0.insert(user_id, Session::new(transmitter)) {
+            error!("Registered already existing session for {user_id}");
+        }
+    }
+    pub async fn close_session(&self, session_id: &Uuid) {
+        if let Some((_, mut session)) = self.0.remove(session_id) {
+            trace!("Sending closing message to session {session_id}");
+            session.close().await.unwrap();
```
Comment (ohaddahan) [inline libs/block-mesh-manager/src/ws/connection_manager.rs:25]: `unwrap` goes boom.
use `?` with `anyhow::Restul`.

File: libs/block-mesh-manager/src/ws/connection_manager.rs [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+
+impl Session {
+    fn new(transmitter: SplitSink<WebSocket, Message>) -> Self {
+        Self { transmitter }
+    }
+    async fn close(&mut self) -> Result<(), axum::Error> {
+        self.transmitter
+            .send(Message::Text(format!("close")))
+            .await?;
+        let close = self.transmitter.close().await.unwrap(); // waits for the sink to close
```
Comment (ohaddahan) [inline libs/block-mesh-manager/src/ws/connection_manager.rs:53]: `unwrap`

File: libs/block-mesh-common/src/interfaces/ws_api.rs [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub enum WsMessageTypes {
+    SendTaskFromServer(GetTaskResponse),
+    SubmitTaskToServer(SubmitTaskRequest),
+    SendBandwidthReportFromServer,
+    SubmitForBandwidthReportToServer(ReportBandwidthRequest),
+    SendUptimeFromServer,
+    SubmitUptimeToServer(ReportUptimeRequest),
+}
```
Comment (ohaddahan) [inline libs/block-mesh-common/src/interfaces/ws_api.rs:25]: Need to remove.

File: libs/block-mesh-common/src/interfaces/ws_api.rs [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+use uuid::Uuid;
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct WsMessage {
+    // only for requests from client
+    pub message_id: Uuid,
+    pub email: Option<String>,
+    pub device: Option<DeviceType>,
+    pub message: WsMessageTypes,
+}
```
Comment (ohaddahan) [inline libs/block-mesh-common/src/interfaces/ws_api.rs:15]: Still relevant?

File: libs/block-mesh-manager/src/worker/ws_worker.rs (hunk 8 above)
Comment (ohaddahan) [reply libs/block-mesh-manager/src/worker/ws_worker.rs:1]: Delete if not needed

File: libs/block-mesh-manager/src/ws/connection_manager.rs [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+use tokio::sync::broadcast;
+use tokio::sync::broadcast::error::SendError;
+use tokio::task::{JoinHandle, JoinSet};
+use tracing::{error, trace, warn};
+use uuid::Uuid;
+
+#[derive(Debug, Clone)]
+pub struct ConnectionManager {
+    pub broadcaster: Broadcaster,
+    pub task_scheduler: TaskScheduler<GetTaskResponse>,
```
Comment (ohaddahan) [inline libs/block-mesh-manager/src/ws/connection_manager.rs:21]: Should be `WsServerMessage`.

File: libs/block-mesh-manager/src/ws/connection_manager.rs [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+        // demo
+        let broadcast_handle = tokio::spawn(async move {
+            loop {
+                tracing::info!("Sending demo broadcast");
+                println!("Sending demo broadcast");
+                if tx.send(WsServerMessage::RequestBandwidthReport).is_err() {
+                    tokio::time::sleep(Duration::from_secs(10)).await;
+                }
+                tokio::time::sleep(Duration::from_secs(1)).await;
+            }
```
Comment (ohaddahan) [inline libs/block-mesh-manager/src/ws/connection_manager.rs:51]: remove debug

File: libs/block-mesh-manager/src/ws/connection_manager.rs [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+}
+#[derive(Debug, Clone)]
+pub struct Broadcaster {
+    transmitter: broadcast::Sender<WsServerMessage>,
+    sockets: Arc<DashMap<Uuid, tokio::sync::mpsc::Sender<WsServerMessage>>>,
+}
+
+impl Broadcaster {
+    fn new() -> Self {
+        let (transmitter, _) = broadcast::channel(10000);
```
Comment (ohaddahan) [inline libs/block-mesh-manager/src/ws/connection_manager.rs:39]: If we aren't using the `rx` why do we need this?
It transmit to no one?

Comment (ohaddahan) [reply libs/block-mesh-manager/src/ws/connection_manager.rs:39]: ahh ok we use `transmitter.subscribe` to generate `rx`

File: libs/block-mesh-manager/src/ws/handle_socket.rs [hunk 16]
```diff
@@ -0,0 +1,134 @@
...
+
+/// Actual websocket statemachine (one will be spawned per connection)
+pub async fn handle_socket(
+    socket: WebSocket,
+    who: SocketAddr,
+    state: Arc<AppState>,
+    _email: String,
+    user_id: Uuid,
+) {
+    let is_closing = Arc::new(AtomicBool::new(false));
```
Comment (ohaddahan) [inline libs/block-mesh-manager/src/ws/handle_socket.rs:21]: Why Arc wrapping an Atomic?

**Pull request merged**